src/api/bdaddr.rs

```diff
@@ -0,0 +1,270 @@
+    type Error = ParseBDAddrError;
+
+    fn try_from(slice: &'a [u8]) -> Result<Self, Self::Error> {
+        if slice.len() < 6 {
+            Err(ParseBDAddrError::IncorrectByteCount)
+        } else {
+            let mut cpy = [0; 6];
+            cpy.copy_from_slice(&slice[..6]);
+            Ok(cpy.into())
+        }
+    }
+}
+
+impl From<u64> for BDAddr {
+    fn from(int: u64) -> Self {
+        let mut cpy = [0; 6];
+        let slice = int.to_be_bytes(); // Reverse order to have MSB on index 0
+        cpy.copy_from_slice(&slice[2..]);
+        cpy.into()
+    }
+}
+
+impl From<BDAddr> for u64 {
+    fn from(addr: BDAddr) -> Self {
+        let mut slice = [0; 8];
+        (&mut slice[2..]).copy_from_slice(&addr.into_inner());
+        u64::from_be_bytes(slice)
+    }
+}
+
+impl From<ParseBDAddrError> for Error {
+    fn from(e: ParseBDAddrError) -> Self {
+        Error::Other(format!("ParseBDAddrError: {}", e))
+    }
+}
+
+impl FromStr for BDAddr {
+    type Err = ParseBDAddrError;
+
+    /// Parses a Bluetooth address of the form `aa:bb:cc:dd:ee:ff` or of form
+    /// `aabbccddeeff`.
+    ///
+    /// All hex-digits `[0-9a-fA-F]` are allowed.
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        if s.contains(':') {
+            Self::from_str_delim(s)
+        } else {
+            Self::from_str_no_delim(s)
+        }
+    }
+}
+
+impl BDAddr {
+    pub fn into_inner(self) -> [u8; 6] {
+        self.address
+    }
+    pub fn bytes(&self) -> &[u8; 6] {
+        &self.address
+    }
+    /// Check if this address is a randomly generated.
+    pub fn is_random_static(&self) -> bool {
+        self.address[5] & 0b11 == 0b11
+    }
+    /// Parses a Bluetooth address colons `:` as delimiters.
+    ///
+    /// All hex-digits `[0-9a-fA-F]` are allowed.
+    pub fn from_str_delim(s: &str) -> Result<Self, ParseBDAddrError> {
+        let bytes = s
+            .split(':')
+            .map(|part: &str| {
+                u8::from_str_radix(part, 16).map_err(|_| ParseBDAddrError::InvalidDigit)
+            })
+            .collect::<Result<Vec<u8>, _>>()?;
+
+        if bytes.len() == 6 {
+            let mut address = [0; 6];
+            address.copy_from_slice(bytes.as_slice());
+            Ok(BDAddr { address })
+        } else {
+            Err(ParseBDAddrError::IncorrectByteCount)
+        }
+    }
+    /// Parses a Bluetooth address without delimiters.
+    ///
+    /// All hex-digits `[0-9a-fA-F]` are allowed.
+    pub fn from_str_no_delim(s: &str) -> Result<Self, ParseBDAddrError> {
+        if s.len() != 12 {
+            return Err(ParseBDAddrError::IncorrectByteCount);
+        }
+        if s.bytes().any(|b| !b.is_ascii_hexdigit()) {
+            return Err(ParseBDAddrError::InvalidDigit);
+        }
+
+        let mut address = [0; 6];
+        for i in (0..12).step_by(2) {
+            let part = &s[i..i + 2];
+            address[i / 2] = u8::from_str_radix(part, 16).expect("Checked upfront");
```

Comment:
Rather than checking twice, remove the `is_ascii_hexdigit` check above, and return the error here instead.